src/webui/api/torrentscontroller.cpp

```diff
@@ -527,15 +626,83 @@ void TorrentsController::addTrackersAction()
     const QString hash = params()["hash"];
 
     BitTorrent::TorrentHandle *const torrent = BitTorrent::Session::instance()->findTorrent(hash);
-    if (torrent) {
-        QList<BitTorrent::TrackerEntry> trackers;
-        foreach (QString url, params()["urls"].split('\n')) {
-            url = url.trimmed();
-            if (!url.isEmpty())
-                trackers << url;
+    if (!torrent)
+        throw APIError(APIErrorType::NotFound);
+
+    QList<BitTorrent::TrackerEntry> trackers;
+    for (QString urlStr : params()["urls"].split('\n')) {
+        urlStr = urlStr.trimmed();
+        QUrl url(urlStr);
+        if (!urlStr.isEmpty() && url.isValid())
+            trackers << urlStr;
+    }
+    torrent->addTrackers(trackers);
+}
+
+void TorrentsController::editTrackerAction()
```

Comment:
Why do you need "edit tracker" abstraction? Since there are no more properties of the tracker that determine its individuality among other trackers of the same torrent, it is enough to have only "add" and "remove" actions.
Let's compare set of integers and set of objects (having several properties). In tge first case we can just remove one int and add another (like in your case), but in the second case we should have separate "edit" action since "remove" + "add" can have different side effect.

Reply:
Probably to make the change atomic.

Reply:
>Probably to make the change atomic.

Well, it can be acceptable reason in case of Web API. But the correct name for this action is "replace".
The more atomic action is replace  current set of trackers with the new one.

Reply:
> But the correct name for this action is "replace".

This is an implementation detail that shouldn't be exposed to the user. To them, they're "editing" a specific tracker, which also aligns with the nomenclature in the GUI.

Reply:
The API is exposed to users as well...